packages/langium/src/service-registry.ts

```diff
@@ -4,38 +4,41 @@
  * terms of the MIT License, which is available in the project root.
  ******************************************************************************/
 
-import type { LangiumServices } from './services.js';
+import type { LangiumCoreServices } from './services.js';
 import { UriUtils, type URI } from './utils/uri-utils.js';
 
 /**
- * The service registry provides access to the language-specific services. These are resolved
- * via the URI of a text document.
+ * The service registry provides access to the language-specific {@link LangiumCoreServices} optionally including LSP-related services.
+ * These are resolved via the URI of a text document.
  */
 export interface ServiceRegistry {
 
     /**
      * Register a language via its injected services.
      */
-    register(language: LangiumServices): void;
+    register(language: LangiumCoreServices): void;
 
     /**
      * Retrieve the language-specific services for the given URI. In case only one language is
      * registered, it may be used regardless of the URI format.
      */
-    getServices(uri: URI): LangiumServices;
+    getServices(uri: URI): LangiumCoreServices;
```

Comment:
How can I access the LSP services if this method only returns `LangiumCoreServices`?

I'm currently trying to run two language servers in the same document, one of which is the Langium language server and the other not.
The Langium language server is the default one, and if the user typed some specific text, it would use the other language server suggestions.

Anyway, the approach I'm taking is to create two modes; according to mode, the language server would use either the language server or the other language server methods to suggest, but now I can't use the language server since I can't directly access the LSP services.


Here is an example:

```typescript
export class LangiumMode implements LanguageMode {
  private getServices({ uri }: TextDocument): LangiumServices /* error in 3.0.0 */ { 
    return this.shared.ServiceRegistry.getServices(URI.parse(uri));
  }

  public doComplete(
    document: TextDocument,
    params: CompletionParams,
    cancelToken?: CancellationToken | undefined,
  ): MaybePromise<CompletionList | undefined> {
    return this.getServices(document).lsp.CompletionProvider?.getCompletion(
      this.textDocumentToLangiumDocument(document),
      params,
      cancelToken,
    );
  }
  
  // ...
}
```

Reply:
for me it worked to downcast

Reply:
@Yokozuna59  A quick answer: Look https://github.com/eclipse-langium/langium/issues/1166#issuecomment-1945711393 and below.

Reply:
I see, thanks @cdietrich @sailingKieler.